### ISSUE TITLE
Problems with serializing nested `Valid` and `Invalid`

### ISSUE THREAD

**Issue by TimSoethout** [romix/akka-kryo-serialization#100]
I'm using `ValidatedNel` from the `cats` package. I'm experiencing the following issue:
When akka tries to serialise my case class which contains a nested `ValidatedNel` it will not work. It seems to get the nesting wrong for the names of the fields or something. I crudely managed to reproduce it by hacking the unit tests a bit.

See here: https://github.com/TimSoethout/akka-kryo-serialization/commit/a87108e72e4921f615bb42f8024e76957f2b47fe

When you run the unit test it breaks hard:
```
> sbt "testOnly com.romix.akka.serialization.kryo.SnapshotRecoveryTest"
[info] Loading global plugins from /Users/tim/.sbt/0.13/plugins
[info] Loading project definition from /Users/tim/Downloads/akka-kryo-serialization/project
[info] Set current project to akka-kryo-serialization (in build file:/Users/tim/Downloads/akka-kryo-serialization/)
[info] Compiling 1 Scala source to /Users/tim/Downloads/akka-kryo-serialization/target/scala-2.11/test-classes...
[WARN] [10/11/2016 15:07:51.091] [pool-6-thread-7-ScalaTest-running-SnapshotRecoveryTest] [akka.serialization.Serialization(akka://testSystem)] Multiple serializers found for class cats.data.Validated$Valid, choosing first: Vector((interface scala.Product,com.romix.akka.serialization.kryo.KryoSerializer@219462dc), (interface java.io.Serializable,akka.serialization.JavaSerializer@4540d92a))
Implicitly registered class with id: cats.data.Validated$Valid=1209028387
Implicitly registered class with id: cats.data.Validated$Invalid=35569944
[WARN] [10/11/2016 15:07:51.278] [pool-6-thread-7-ScalaTest-running-SnapshotRecoveryTest] [akka.serialization.Serialization(akka://testSystem)] Multiple serializers found for class cats.data.Validated$Invalid, choosing first: Vector((interface scala.Product,com.romix.akka.serialization.kryo.KryoSerializer@219462dc), (interface java.io.Serializable,akka.serialization.JavaSerializer@4540d92a))
Implicitly registered class with id: cats.data.NonEmptyList=413260567
Implicitly registered class with id: scala.collection.immutable.Nil$=2013895718
Success([B@8fe4fdc)
Success([B@695eb0e4)
[WARN] [10/11/2016 15:07:51.286] [pool-6-thread-7-ScalaTest-running-SnapshotRecoveryTest] [akka.serialization.Serialization(akka://testSystem)] Multiple serializers found for class scala.Some, choosing first: Vector((interface scala.Product,com.romix.akka.serialization.kryo.KryoSerializer@219462dc), (interface java.io.Serializable,akka.serialization.JavaSerializer@4540d92a))
Implicitly registered class with id: scala.Some=1288426183
Success([B@57e30b80)
Success([B@4ba65808)
[WARN] [10/11/2016 15:07:51.287] [pool-6-thread-7-ScalaTest-running-SnapshotRecoveryTest] [akka.serialization.Serialization(akka://testSystem)] Multiple serializers found for class com.romix.akka.serialization.kryo.Test, choosing first: Vector((interface scala.Product,com.romix.akka.serialization.kryo.KryoSerializer@219462dc), (interface java.io.Serializable,akka.serialization.JavaSerializer@4540d92a))
Implicitly registered class with id: com.romix.akka.serialization.kryo.Test=662069142
[WARN] [10/11/2016 15:07:51.288] [pool-6-thread-7-ScalaTest-running-SnapshotRecoveryTest] [akka.serialization.Serialization(akka://testSystem)] Multiple serializers found for class com.romix.akka.serialization.kryo.ValidatedWrapper, choosing first: Vector((interface scala.Product,com.romix.akka.serialization.kryo.KryoSerializer@219462dc), (interface java.io.Serializable,akka.serialization.JavaSerializer@4540d92a))
Implicitly registered class with id: com.romix.akka.serialization.kryo.ValidatedWrapper=415019382
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000101f786d8, pid=98382, tid=0x0000000000003d0b
#
# JRE version: Java(TM) SE Runtime Environment (8.0_92-b14) (build 1.8.0_92-b14)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.92-b14 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.dylib+0x5786d8]  Unsafe_GetObject+0x51
#
# Core dump written. Default location: /cores/core or core.98382
#
# An error report file with more information is saved as:
# /Users/tim/Downloads/akka-kryo-serialization/hs_err_pid98382.log
Compiled method (nm)   12510 15493     n 0       sun.misc.Unsafe::getObject (native)
 total in heap  [0x0000000104069150,0x00000001040694a8] = 856
 relocation     [0x0000000104069278,0x00000001040692b8] = 64
 main code      [0x00000001040692c0,0x00000001040694a8] = 488
Compiled method (nm)   12510 15493     n 0       sun.misc.Unsafe::getObject (native)
 total in heap  [0x0000000104069150,0x00000001040694a8] = 856
 relocation     [0x0000000104069278,0x00000001040692b8] = 64
 main code      [0x00000001040692c0,0x00000001040694a8] = 488
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
/usr/local/Cellar/sbt/0.13.11/libexec/sbt-launch-lib.bash: line 41: 98382 Abort trap: 6           (core dumped) "$@"
```

Relates:
When I try something like this in my own project in the unit tests (borrowing your `roundTrip` from `ScalaKryoTest` :
```scala

case class Test(s:String)
case class ValidatedWrapper(v: ValidatedNel[String, Int])

"BIG CRASH" should "not crash" in {

    val valid: ValidatedNel[String, Int] = Validated.valid(1)
    val invalid: ValidatedNel[String, Int] = Validated.invalidNel("errrr")

    val v = roundTrip(valid)
    val iv = roundTrip(invalid)

    println(v)
    println(iv)

    val wrappedv: Option[ValidatedNel[String, Int]] = Some(valid)
    val wrappediv: Option[ValidatedNel[String, Int]] = Some(valid)

    val wv = roundTrip(wrappedv)
    val wiv = roundTrip(wrappediv)


    println(wv)
    println(wiv)


    roundTrip(Test("test"))

    roundTrip(ValidatedWrapper(valid))
    roundTrip(ValidatedWrapper(invalid))

  }
```

it does not crash but gives me NPE:

```
Implicitly registered class with id: cats.data.Validated$Valid=1209028387
Implicitly registered class with id: cats.data.Validated$Invalid=35569944
Implicitly registered class with id: cats.data.NonEmptyList=413260567
Implicitly registered class with id: scala.collection.immutable.Nil$=2013895718
Valid(1)
Invalid(NonEmptyList(errrr))
Implicitly registered class with id: scala.Some=1288426183
Some(Valid(1))
Some(Valid(1))
Implicitly registered class with id: com.ing.rebel.kryo.Test=344068600
Implicitly registered class with id: com.ing.rebel.kryo.ValidatedWrapper=2073457946

java.lang.NullPointerException
Serialization trace:
head (java.lang.Integer)
a (cats.data.Validated$Valid)
v (com.ing.rebel.kryo.ValidatedWrapper)
com.esotericsoftware.kryo.KryoException: java.lang.NullPointerException
Serialization trace:
head (java.lang.Integer)
a (cats.data.Validated$Valid)
v (com.ing.rebel.kryo.ValidatedWrapper)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:101)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:518)
	at com.esotericsoftware.kryo.Kryo.writeObjectOrNull(Kryo.java:606)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:87)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:518)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:552)
	at com.esotericsoftware.kryo.serializers.ObjectField.write(ObjectField.java:80)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:518)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:628)
	at xxxx.kryo.SpecCase.roundTrip(SpecCase.scala:39)
	at xxxx.kryo.ScalaKryoTest$$anonfun$3.apply(ScalaKryoTest.scala:56)
	at xxxx.kryo.ScalaKryoTest$$anonfun$3.apply(ScalaKryoTest.scala:27)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
	at xxxx.kryo.ScalaKryoTest.withFixture(ScalaKryoTest.scala:78)
	at org.scalatest.FlatSpecLike$class.invokeWithFixture$1(FlatSpecLike.scala:1679)

```

**Comment by TimSoethout** [comment]
It does seem to work with `Either`:

```scala
  case class EitherWrapper(v: Either[String, Int])
  ...

  it should "handle nested Either" in {
    val left: Either[String, Int] = Left("s")
    val right: Either[String, Int] = Right(1)

    roundTrip(EitherWrapper(left))
    roundTrip(EitherWrapper(right))
  }
```

results in
```
Testing started at 16:06 ...
Implicitly registered class with id: com.ing.rebel.kryo.EitherWrapper=8927732
Implicitly registered class with id: scala.util.Left=946164401
Implicitly registered class with id: scala.util.Right=199227405

Process finished with exit code 0
```

**Comment by ktoso** [comment]
I can reproduce and confirm this issue, however I believe the problem is in core Kryo.
Debugged it a while was not able to locate the issue so far.

Interesting observation is that running with TRACE logging (via `Log.TRACE()`) makes it serialize the message properly, see below:


I prepared a full reproducer, however believe this should be fixed in Kryo core instead of this plugin so far.

Full reproducer with outputs: https://github.com/ktoso/akka-cats-serialize/tree/master
Both successful run with TRACE and failing run without logging

**Comment by luben** [comment]
I think this issue is addressed in Kryo-4.0 that changes how the generic classes are serialized by the FieldsSerializer that is used here. I have to test and confirm. Thanks @ktoso for the isolated test case

**Comment by TimSoethout** [comment]
Is there an easy way to setup 4.0 for akka?

**Comment by luben** [comment]
You can try to change the dependency to kryo-4.0.0, recompile and publish it locally. I don't think there is need for any change.

**Comment by ktoso** [comment]
I confirm that Kryo `4.0.0` resolves this issue.
It also seems to be drop-in compatible with this version of akka-kryo-serialization.

I don't know about what timing would be best for this library to upgrade to Kryo 4.0, but it would be great to do so soon it seems :-)

@TimSoethout in order to make this work for you right away, you can override the kryo dependency, as shown here: https://github.com/ktoso/akka-cats-serialize/commit/3e95dcdaa3b81c3f66104392454046a1eff7d9b8

**Comment by magro** [comment]
To allow clients to retain the former handling of generics a config option to control `kryo.getFieldSerializerConfig().setOptimizedGenerics(true)` could be added.

**Comment by TimSoethout** [comment]
Cheers, thanks @ktoso, @luben & @magro. This seems to solve it for me for now. 

I do run into another issue where I nest a unit in a `Validation`, but that one is already documented here: https://github.com/romix/akka-kryo-serialization/issues/89.

**Comment by ktoso** [comment]
Great, thanks everyone involved.

@luben when do you think this plugin could upgrade to 4.0? Seems like a general good step to do in any case in the near future.

**Comment by TimSoethout** [comment]
@magro, As mentioned in #89:
`kryo.getFieldSerializerConfig.setOptimizedGenerics(true)` seems to let the NPE return.
Maybe this was the crux all along?

**Comment by luben** [comment]
Yes, couple of weeks ago I proposed to cut 0.4.2 with the currently commited improvements and kryo-3.0 and also release 0.5.0 depending on kryo-4.0 with settings to turn on the backward compatibility with 3.0 serialization format.  https://github.com/romix/akka-kryo-serialization/pull/97

**Comment by ktoso** [comment]
Great, thanks!

**Comment by romix** [comment]
@luben I'm not opposed to release 0.4.2 and 0.5.0 as you suggested. I'm just too busy at work recently. If you can submit PRs for this, I'm fine reviewing them and rolling out the release.

**Comment by luben** [comment]
@romix great. I will give a shot with the Unit serialization problem and will prepare PR for 0.4.2 release and later 0.5.0 release based on kryo-4.0 (with added configs to keep the backward compatibility with 3.0)

**Comment by romix** [comment]
@luben Thanks a lot!

**Comment by ktoso** [comment]
Awesome, love this community @luben @romix :-)
